crates/iota-data-ingestion-core/src/executor.rs

```diff
@@ -66,41 +82,102 @@ impl<P: ProgressStore> IndexerExecutor<P> {
+                                gc_sender.send(seq_number).await?;
+                                reader_checkpoint_number = seq_number;
+                            }
+                            self.metrics.data_ingestion_checkpoint.with_label_values(&[&task_name]).set(sequence_number as i64);
+                        }
+                        // Manages the graceful shutdown sequence of the entire indexer system.
+                        //
+                        // The shutdown process follows these steps:
+                        // 1. Token cancellation triggers:
+                        //    a. Individual workers in each pool:
+                        //       - Complete current checkpoint processing
+                        //       - Send final progress updates
+                        //       - Signal completion to their pool
+                        //    b. Worker pools:
+                        //       - Stop accepting new checkpoints
+                        //       - Process remaining progress messages
+                        //       - Wait for all their workers to finish
+                        //       - Send ShutDown message to executor
+                        //
+                        // 2. Executor main loop:
+                        //    - Continues processing Progress messages from pools
+                        //    - Tracks pool shutdowns via ShutDown messages
+                        //    - Once all pools report shutdown:
+                        //      a. Awaits all worker pool join handles
+                        //      b. Signals checkpoint reader to stop
+                        //      c. Awaits checkpoint reader completion
+                        //      d. Exits main loop
+                        //
+                        // This ensures hierarchical shutdown order:
+                        // 1. Workers (in parallel within each pool)
+                        // 2. Worker pools (in parallel)
+                        // 3. Checkpoint reader
+                        // 4. Executor main loop
+                        //
+                        // Guarantees:
+                        // - No work is interrupted mid-processing
+                        // - All progress is saved to storage
+                        // - All messages are processed in order
+                        // - All resources are properly cleaned up
+                        WorkerPoolMsg::ShutDown(worker_pool_name) => {
+                            // Track worker pools that have initiated shutdown
+                            worker_pools_shutdown_signals.push(worker_pool_name);
+                            // Once all workers pools have signaled completion, await their handles
+                            // This ensures all workers have finished their final tasks
+                            if worker_pools_shutdown_signals.len() == self.pool_senders.len() {
+                                for worker in worker_pools {
```

Comment:
It is important to distinguish between `worker` and `worker_pool`, as they have different scope and hence shutdown logic.

```rust
                                for worker_pool in worker_pools {
```

Reply:
[fixed](https://github.com/iotaledger/iota/pull/4648/commits/f2d9f8da3328c5aae5815778d69daba4cdbd5bc9#diff-fc30500a6b8c33380223c735cb9d6786afa2cc39301199f614993173f3acdb85R150-R153)